### ISSUE TITLE
NTYPE versus CTYPE

### ISSUE THREAD

**Issue by phoe** [marcoheisig/Typo#1]
Stupid question because I'm randomly going through the code - what's the difference betwen `NTYPE` from this implementation and `CTYPE` from https://github.com/s-expressionists/ctype ?

**Comment by marcoheisig** [comment]
Very good question :)

`CTYPE` is a full-fledged, precise implementation of CL types, with its own versions of `typep` and `subtypep`. It requires some amount of implementation specific hooks to be useful.

`NTYPE` is only does approximate reasoning about types, but is really fast and doesn't cons.  It relies on the host's versions of `typep` and `subtypep` to do the heavy lifting.  But it is faster (which matters for Petalisp), and fully portable.  The main goal of `NTYPE` is not to narrow down the type of each value in a program enough to choose a specialized representation.

The cool thing about Typo is that it can also (portably!) convert s-expressions to the most applicable specialized version applicable to its target types.  For example, it can replace
```lisp
(cl:+ a b c)
```
where `a` is an integer, `b` is a double-float, and `c` is a single-float with
```lisp
(two-arg-f64+ (two-arg-f64+ (f64-from-integer a) b) (f64-from-f32 c))
```
This, in turn, can be used to optimize the code further, or to analyze its performance.  The main use-case for this is Petalisp.  The other intended use-case is Loopus, my loop optimization library.

tl;dr:  I only care about type inference, not about reliably answering `typep` or `subtypep` queries.

**Comment by phoe** [comment]
Thanks, this answers my question! (Would it be good to copypaste this whole post as-is into Typo's README? :D)

**Comment by marcoheisig** [comment]
ERMAGHERD This is a two days old project, and people are already nagging me about documentation :)

I'll add this to the README once things actually work, which may well be today or tomorrow.

**Comment by phoe** [comment]
Okay. :D Thanks.

Feel free to close this issue whenever ~~or let it hang as a reminder of my nagging stare at your README file~~.

**Comment by marcoheisig** [comment]
I updated the README.

**Comment by phoe** [comment]
Thanks!
